flink-runtime/src/main/java/org/apache/flink/streaming/api/functions/source/InputFormatSource.java

```diff
@@ -0,0 +1,314 @@
+
+        @Override
+        public void addReader(int subtaskId) {}
+
+        @Override
+        public Void snapshotState(long checkpointId) {
+            return null;
+        }
+
+        @Override
+        public void close() {}
+    }
+
+    private static final long serialVersionUID = 1L;
+
+    private final Boundedness boundedness;
+    private final InputFormat<OUT, InputSplit> format;
+
+    @SuppressWarnings("unchecked")
+    public InputFormatSource(Boundedness boundedness, InputFormat<OUT, ?> format) {
+        this.boundedness = boundedness;
+        this.format = (InputFormat<OUT, InputSplit>) format;
+    }
+
+    @Override
+    public Boundedness getBoundedness() {
+        return boundedness;
+    }
+
+    @Override
+    public SplitEnumerator<SourceSplit, Void> createEnumerator(
+            SplitEnumeratorContext<SourceSplit> context) throws Exception {
+        return new InputFormatSplitEnumerator<>(format, context);
+    }
+
+    @Override
+    public SplitEnumerator<SourceSplit, Void> restoreEnumerator(
+            SplitEnumeratorContext<SourceSplit> context, Void checkpoint) throws Exception {
+        return new InputFormatSplitEnumerator<>(format, context);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<SourceSplit> getSplitSerializer() {
+        return new SimpleVersionedSerializer<>() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
+
+            @Override
+            public byte[] serialize(SourceSplit split) throws IOException {
+                return InstantiationUtil.serializeObject(split);
+            }
+
+            @Override
+            public SourceSplit deserialize(int version, byte[] serialized) throws IOException {
+                try {
+                    return InstantiationUtil.deserializeObject(
+                            serialized, SourceSplit.class.getClassLoader());
+                } catch (ClassNotFoundException e) {
+                    throw new IOException("Failed to deserialize SourceSplit.");
+                }
+            }
+        };
+    }
+
+    @Override
+    public SimpleVersionedSerializer<Void> getEnumeratorCheckpointSerializer() {
+        return new SimpleVersionedSerializer<>() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
+
+            @Override
+            public byte[] serialize(Void obj) {
+                return new byte[0];
+            }
+
+            @Override
+            public Void deserialize(int version, byte[] serialized) {
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public SourceReader<OUT, SourceSplit> createReader(SourceReaderContext readerContext)
+            throws Exception {
+        Preconditions.checkState(
+                readerContext instanceof RichSourceReaderContext,
+                "Source reader context must be an instance of RichSourceReaderContext");
```

Comment:
What's the point of this check? The user doesn't control this and it is basically checked/enforced on the next line with the cast

Reply:
The point is that if somebody is playing with it it must be obvious what's the original intention. When I see legacy codes I miss such many times which leaves you in doubts whether something is accident or plan 🙂